### PR TITLE
add bootloader HAL functions to the luos HAL library

### PR DESCRIPTION
**Pull request by assembly-lab**

### CODE REVIEW

File: STM32L4/luos_hal_config.h [hunk 1]
```diff
@@ -163,4 +163,15 @@
 #define ADDRESS_LAST_PAGE_FLASH ((uint32_t)(FLASH_END - FLASH_PAGE_SIZE))
 #endif
 
+/*******************************************************************************
+ * BOOTLOADER CONFIG
+ ******************************************************************************/
+#define SHARED_MEMORY_ADDRESS 0x0800C000
+#define APP_ADDRESS           (uint32_t)0x0800C800
+
+#define BOOT_MODE_MASK   0x000000FF
```
Comment (Salem-Tho) [inline STM32L4/luos_hal_config.h]: if this define never change should be in luos_hal.c or .h
is it really thing user want to change?

File: STM32L4/luos_hal.c [hunk 2]
```diff
@@ -21,6 +21,10 @@
 #include "stm32l4xx_ll_dma.h"
 #include "stm32l4xx_ll_system.h"
 
+#ifdef BOOTLOADER_CONFIG
+#include <math.h>
```
Comment (Salem-Tho) [inline STM32L4/luos_hal.c]: why math.h rellay needed?

File: STM32L4/luos_hal.c [hunk 3]
```diff
@@ -745,3 +749,217 @@ void LuosHAL_FlashReadLuosMemoryInfo(uint32_t addr, uint16_t size, uint8_t *data
...
+    NVIC_SystemReset();
+}
+
+#ifdef BOOTLOADER_CONFIG
+/******************************************************************************
+ * @brief Delay - blocking
+ * @param 
+ * @return
+ ******************************************************************************/
+void LuosHAL_Delay(uint32_t delay)
```
Comment (Salem-Tho) [inline STM32L4/luos_hal.c]: use luosHAL_GetSystick in a while loop

File: STM32L4/luos_hal.c [hunk 4]
```diff
@@ -745,3 +749,217 @@ void LuosHAL_FlashReadLuosMemoryInfo(uint32_t addr, uint16_t size, uint8_t *data
 {
     memcpy(data, (void *)(addr), size);
 }
+
+#ifdef BOOTLOADER_CONFIG
```
Comment (Salem-Tho) [inline STM32L4/luos_hal.c]: I prefere to have all the function at the end of the file with only one #ifdef

**Review by Salem-Tho — COMMENTED** [review]
all the note for the L4 are applicable to the other the HAL

**Review by nicolas-rabault — COMMENTED** [review]
sounds good. Agree with @Salem-Luos comments

File: STM32L4/luos_hal_config.h (hunk 1 above)
Comment (assembly-lab) [reply STM32L4/luos_hal_config.h]: Ok, I made these changes.

File: STM32L4/luos_hal.c (hunk 2 above)
Comment (assembly-lab) [reply STM32L4/luos_hal.c]: I deleted these lines.

File: STM32L4/luos_hal.c (hunk 3 above)
Comment (assembly-lab) [reply STM32L4/luos_hal.c]: OK I made these changes.

File: STM32L4/luos_hal.c (hunk 4 above)
Comment (assembly-lab) [reply STM32L4/luos_hal.c]: OK, I made these changes.

**Review by Salem-Tho — APPROVED** [review]

**Pull request merged**